File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/One.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class One {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;

    List<Object> selfEdges = List.of();
    List<Object> aEdges = List.of();
    

    static class OneCleaner implements Runnable {

        MemorySegment segment;
        OneCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.One_destroy(this.segment);
        }
    }

    One() {}
    One(MemorySegment handle, List<Object> selfEdges, List<Object> aEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        this.aEdges = aEdges;
        

    }
    
    public static One transitivity(One hold,One nohold) {
        
        var holdNative = hold.internal;
        var noholdNative = nohold.internal;
        var nativeVal = somelib_h.One_transitivity(holdNative, noholdNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> aEdges = List.of(hold);
        var returnVal = new One(nativeVal, selfEdges, aEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One cycle(Two hold,One nohold) {
        
        var holdNative = hold.internal;
        var noholdNative = nohold.internal;
        var nativeVal = somelib_h.One_cycle(holdNative, noholdNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> aEdges = List.of(hold);
        var returnVal = new One(nativeVal, selfEdges, aEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One manyDependents(One a,One b,Two c,Two d,Two nohold) {
        
        var aNative = a.internal;
        var bNative = b.internal;
        var cNative = c.internal;
        var dNative = d.internal;
        var noholdNative = nohold.internal;
        var nativeVal = somelib_h.One_many_dependents(aNative, bNative, cNative, dNative, noholdNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> aEdges = List.of(a, b, c, d);
        var returnVal = new One(nativeVal, selfEdges, aEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One returnOutlivesParam(Two hold,One nohold) {
        
        var holdNative = hold.internal;
        var noholdNative = nohold.internal;
        var nativeVal = somelib_h.One_return_outlives_param(holdNative, noholdNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> longEdges = List.of(hold);
        var returnVal = new One(nativeVal, selfEdges, longEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One diamondTop(One top,One left,One right,One bottom) {
        
        var topNative = top.internal;
        var leftNative = left.internal;
        var rightNative = right.internal;
        var bottomNative = bottom.internal;
        var nativeVal = somelib_h.One_diamond_top(topNative, leftNative, rightNative, bottomNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> topEdges = List.of(top, left, right, bottom);
        var returnVal = new One(nativeVal, selfEdges, topEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One diamondLeft(One top,One left,One right,One bottom) {
        
        var topNative = top.internal;
        var leftNative = left.internal;
        var rightNative = right.internal;
        var bottomNative = bottom.internal;
        var nativeVal = somelib_h.One_diamond_left(topNative, leftNative, rightNative, bottomNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> leftEdges = List.of(left, bottom);
        var returnVal = new One(nativeVal, selfEdges, leftEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One diamondRight(One top,One left,One right,One bottom) {
        
        var topNative = top.internal;
        var leftNative = left.internal;
        var rightNative = right.internal;
        var bottomNative = bottom.internal;
        var nativeVal = somelib_h.One_diamond_right(topNative, leftNative, rightNative, bottomNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> rightEdges = List.of(right, bottom);
        var returnVal = new One(nativeVal, selfEdges, rightEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One diamondBottom(One top,One left,One right,One bottom) {
        
        var topNative = top.internal;
        var leftNative = left.internal;
        var rightNative = right.internal;
        var bottomNative = bottom.internal;
        var nativeVal = somelib_h.One_diamond_bottom(topNative, leftNative, rightNative, bottomNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> bottomEdges = List.of(bottom);
        var returnVal = new One(nativeVal, selfEdges, bottomEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One diamondAndNestedTypes(One a,One b,One c,One d,One nohold) {
        
        var aNative = a.internal;
        var bNative = b.internal;
        var cNative = c.internal;
        var dNative = d.internal;
        var noholdNative = nohold.internal;
        var nativeVal = somelib_h.One_diamond_and_nested_types(aNative, bNative, cNative, dNative, noholdNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> aEdges = List.of(a, b, c, d);
        var returnVal = new One(nativeVal, selfEdges, aEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One implicitBounds(One explicitHold,One implicitHold,One nohold) {
        
        var explicitHoldNative = explicitHold.internal;
        var implicitHoldNative = implicitHold.internal;
        var noholdNative = nohold.internal;
        var nativeVal = somelib_h.One_implicit_bounds(explicitHoldNative, implicitHoldNative, noholdNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> aEdges = List.of(explicitHold, implicitHold);
        var returnVal = new One(nativeVal, selfEdges, aEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    public static One implicitBoundsDeep(One explicit,One implicit1,One implicit2,One nohold) {
        
        var explicitNative = explicit.internal;
        var implicit1Native = implicit1.internal;
        var implicit2Native = implicit2.internal;
        var noholdNative = nohold.internal;
        var nativeVal = somelib_h.One_implicit_bounds_deep(explicitNative, implicit1Native, implicit2Native, noholdNative);
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> aEdges = List.of(explicit, implicit1, implicit2);
        var returnVal = new One(nativeVal, selfEdges, aEdges);
        var cleaner = new One.OneCleaner(nativeVal);
        returnVal.cleanable = Lib.cleaner.register(returnVal, cleaner);
        return returnVal;
    }
    
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/One.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface OneLib: Library {
    fun One_destroy(handle: Pointer)
    fun One_transitivity(hold: Pointer, nohold: Pointer): Pointer
    fun One_cycle(hold: Pointer, nohold: Pointer): Pointer
    fun One_many_dependents(a: Pointer, b: Pointer, c: Pointer, d: Pointer, nohold: Pointer): Pointer
    fun One_return_outlives_param(hold: Pointer, nohold: Pointer): Pointer
    fun One_diamond_top(top: Pointer, left: Pointer, right: Pointer, bottom: Pointer): Pointer
    fun One_diamond_left(top: Pointer, left: Pointer, right: Pointer, bottom: Pointer): Pointer
    fun One_diamond_right(top: Pointer, left: Pointer, right: Pointer, bottom: Pointer): Pointer
    fun One_diamond_bottom(top: Pointer, left: Pointer, right: Pointer, bottom: Pointer): Pointer
    fun One_diamond_and_nested_types(a: Pointer, b: Pointer, c: Pointer, d: Pointer, nohold: Pointer): Pointer
    fun One_implicit_bounds(explicitHold: Pointer, implicitHold: Pointer, nohold: Pointer): Pointer
    fun One_implicit_bounds_deep(explicit: Pointer, implicit1: Pointer, implicit2: Pointer, nohold: Pointer): Pointer
}

class One internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal val aEdges: List<Any?>,
)  {

    internal class OneCleaner(val handle: Pointer, val lib: OneLib) : Runnable {
        override fun run() {
            lib.One_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<OneLib> = OneLib::class.java
        internal val lib: OneLib = Native.load("diplomat_feature_tests", libClass)
        @JvmStatic
        
        fun transitivity(hold: One, nohold: One): One {
            // This lifetime edge depends on lifetimes: 'a, 'b, 'c, 'd, 'e
            val aEdges: MutableList<Any> = mutableListOf(hold);
            
            val returnVal = lib.One_transitivity(hold.handle, nohold.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun cycle(hold: Two, nohold: One): One {
            // This lifetime edge depends on lifetimes: 'a, 'b, 'c
            val aEdges: MutableList<Any> = mutableListOf(hold);
            
            val returnVal = lib.One_cycle(hold.handle, nohold.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun manyDependents(a: One, b: One, c: Two, d: Two, nohold: Two): One {
            // This lifetime edge depends on lifetimes: 'a, 'b, 'c, 'd
            val aEdges: MutableList<Any> = mutableListOf(a, b, c, d);
            
            val returnVal = lib.One_many_dependents(a.handle, b.handle, c.handle, d.handle, nohold.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun returnOutlivesParam(hold: Two, nohold: One): One {
            // This lifetime edge depends on lifetimes: 'long
            val longEdges: MutableList<Any> = mutableListOf(hold);
            
            val returnVal = lib.One_return_outlives_param(hold.handle, nohold.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, longEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun diamondTop(top: One, left: One, right: One, bottom: One): One {
            // This lifetime edge depends on lifetimes: 'top, 'left, 'right, 'bottom
            val topEdges: MutableList<Any> = mutableListOf(top, left, right, bottom);
            
            val returnVal = lib.One_diamond_top(top.handle, left.handle, right.handle, bottom.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, topEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun diamondLeft(top: One, left: One, right: One, bottom: One): One {
            // This lifetime edge depends on lifetimes: 'left, 'bottom
            val leftEdges: MutableList<Any> = mutableListOf(left, bottom);
            
            val returnVal = lib.One_diamond_left(top.handle, left.handle, right.handle, bottom.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, leftEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun diamondRight(top: One, left: One, right: One, bottom: One): One {
            // This lifetime edge depends on lifetimes: 'right, 'bottom
            val rightEdges: MutableList<Any> = mutableListOf(right, bottom);
            
            val returnVal = lib.One_diamond_right(top.handle, left.handle, right.handle, bottom.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, rightEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun diamondBottom(top: One, left: One, right: One, bottom: One): One {
            // This lifetime edge depends on lifetimes: 'bottom
            val bottomEdges: MutableList<Any> = mutableListOf(bottom);
            
            val returnVal = lib.One_diamond_bottom(top.handle, left.handle, right.handle, bottom.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, bottomEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun diamondAndNestedTypes(a: One, b: One, c: One, d: One, nohold: One): One {
            // This lifetime edge depends on lifetimes: 'a, 'b, 'c, 'd
            val aEdges: MutableList<Any> = mutableListOf(a, b, c, d);
            
            val returnVal = lib.One_diamond_and_nested_types(a.handle, b.handle, c.handle, d.handle, nohold.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun implicitBounds(explicitHold: One, implicitHold: One, nohold: One): One {
            // This lifetime edge depends on lifetimes: 'a, 'b, 'c, 'd, 'x
            val aEdges: MutableList<Any> = mutableListOf(explicitHold, implicitHold);
            
            val returnVal = lib.One_implicit_bounds(explicitHold.handle, implicitHold.handle, nohold.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
        @JvmStatic
        
        fun implicitBoundsDeep(explicit: One, implicit1: One, implicit2: One, nohold: One): One {
            // This lifetime edge depends on lifetimes: 'a, 'b, 'c, 'd
            val aEdges: MutableList<Any> = mutableListOf(explicit, implicit1, implicit2);
            
            val returnVal = lib.One_implicit_bounds_deep(explicit.handle, implicit1.handle, implicit2.handle, nohold.handle);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = One(handle, selfEdges, aEdges)
            CLEANER.register(returnOpaque, One.OneCleaner(handle, One.lib));
            return returnOpaque
        }
    }

}